File: GhostChat-iOS/GhostChat/Core/Crypto/IdentityKeyService.swift
import Foundation
import CryptoKit

/// Управляет постоянным P-256 Identity Key пользователя.
/// Генерируется один раз при первом запуске, хранится в Keychain.
/// Этот ключ идентифицирует пользователя между сессиями.
final class IdentityKeyService {

    static let shared = IdentityKeyService()

    private static let keychainKey = "ghost_identity_key"

    private var _privateKey: P256.KeyAgreement.PrivateKey?

    // MARK: - Public API

    /// Private identity key (lazy-loaded from Keychain or generated)
    var privateKey: P256.KeyAgreement.PrivateKey {
        if let key = _privateKey { return key }
        let key = loadOrGenerate()
        _privateKey = key
        return key
    }

    /// Public identity key
    var publicKey: P256.KeyAgreement.PublicKey {
        privateKey.publicKey
    }

    /// x963 representation (65 bytes) for storage in Contact.identityKey
    var publicKeyData: Data {
        publicKey.x963Representation
    }

    /// Base64-encoded x963 public key for key-exchange wire protocol
    func exportPublicKey() -> String {
        publicKeyData.base64EncodedString()
    }

    /// Destroy identity key — panic button (wipes from Keychain + memory)
    func destroy() {
        _privateKey = nil
        KeychainService.delete(forKey: Self.keychainKey)
    }

    // MARK: - Internal

    private func loadOrGenerate() -> P256.KeyAgreement.PrivateKey {
        // Try loading from Keychain
        if let data = KeychainService.load(forKey: Self.keychainKey),
           let key = try? P256.KeyAgreement.PrivateKey(rawRepresentation: data) {
            return key
        }

        // Generate new key
        let key = P256.KeyAgreement.PrivateKey()
        KeychainService.save(key.rawRepresentation, forKey: Self.keychainKey)
        return key
    }

    private init() {}
}

File: GhostChat-Android/app/src/main/java/com/ghost/chat/core/crypto/IdentityKeyService.kt
package com.ghost.chat.core.crypto

import android.util.Base64
import com.ghost.chat.core.security.KeystoreService
import java.security.KeyFactory
import java.security.KeyPair
import java.security.interfaces.ECPublicKey
import java.security.spec.PKCS8EncodedKeySpec

/// Manages the persistent P-256 Identity Key.
/// Generated once on first launch, stored in EncryptedSharedPreferences.
/// This key identifies the user between sessions.
object IdentityKeyService {

    private const val KEYSTORE_KEY_PRIVATE = "ghost_identity_key_private"
    private const val KEYSTORE_KEY_PUBLIC = "ghost_identity_key_public"

    private var _keyPair: KeyPair? = null

    /** Private identity key pair (lazy-loaded from Keystore or generated) */
    val keyPair: KeyPair
        get() {
            _keyPair?.let { return it }
            val kp = loadOrGenerate()
            _keyPair = kp
            return kp
        }

    /** Public identity key */
    val publicKey: ECPublicKey
        get() = keyPair.public as ECPublicKey

    /** x963 representation (65 bytes) for storage in Contact.identityKey */
    val publicKeyData: ByteArray
        get() = DoubleRatchet.exportPublicKeyX963(publicKey)

    /** Base64-encoded x963 public key for key-exchange wire protocol */
    fun exportPublicKey(): String {
        return Base64.encodeToString(publicKeyData, Base64.NO_WRAP)
    }

    /** Destroy identity key — panic button (wipes from Keystore + memory) */
    fun destroy() {
        _keyPair = null
        KeystoreService.delete(KEYSTORE_KEY_PRIVATE)
        KeystoreService.delete(KEYSTORE_KEY_PUBLIC)
    }

    // MARK: - Internal

    private fun loadOrGenerate(): KeyPair {
        // Try loading from Keystore
        val privateKeyData = KeystoreService.load(KEYSTORE_KEY_PRIVATE)
        val publicKeyData = KeystoreService.load(KEYSTORE_KEY_PUBLIC)

        if (privateKeyData != null && publicKeyData != null) {
            try {
                val kf = KeyFactory.getInstance("EC")
                val privateKey = kf.generatePrivate(PKCS8EncodedKeySpec(privateKeyData))
                val publicKey = DoubleRatchet.importPublicKeyX963(publicKeyData)
                return KeyPair(publicKey, privateKey)
            } catch (e: Exception) {
                // Corrupted — regenerate
            }
        }

        // Generate new key
        val kp = DoubleRatchet.generateKeyPair()
        KeystoreService.save(kp.private.encoded, KEYSTORE_KEY_PRIVATE)
        KeystoreService.save(DoubleRatchet.exportPublicKeyX963(kp.public as ECPublicKey), KEYSTORE_KEY_PUBLIC)
        return kp
    }
}
